### PR TITLE
De-duplicate tags between contexts

### PR DESCRIPTION
**Pull request by vickenty**
### What does this PR do?

De-duplicate tags between aggregator contexts. Compute a partial context key (named `TagsKey`) from just the tags, and store the tags in a separate map. String slices are then shared between contexts where possible.

### Motivation

Reduce memory usage for the agent, when multiple metrics are reported with the same tags.

### Additional Notes

This PR moves tags slice from the Context struct to a store Entry. This reduces the size of Context allocation from 64 to 48 bytes, offsetting some of the cost of having the tags store.

### Possible Drawbacks / Trade-offs

We spend additional memory to try and find duplicates in the tagsets. Of course, if there is none (or almost none), we end up allocating more memory to the map than saving from duplicated string slices.

### Describe how to test/QA your changes

Run the agent.
Send metrics via dogstatsd, check that the tags in the app match the tags sent.
Send metrics from a check, check that the tags in the app again.
Stop sending metrics. Check that the contexts and tags are cleaned up and memory is released (via telemetry or heap dump).

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/aggregator/context_resolver.go [hunk 1]
```diff
@@ -9,51 +9,59 @@ import (
...
 func (cr *contextResolver) trackContext(metricSampleContext metrics.MetricSampleContext) ckey.ContextKey {
-	metricSampleContext.GetTags(cr.tagsBuffer)               // tags here are not sorted and can contain duplicates
-	contextKey := cr.generateContextKey(metricSampleContext) // the generator will remove duplicates from cr.tagsBuffer (and doesn't mind the order)
+	metricSampleContext.GetTags(cr.tagsBuffer)                        // tags here are not sorted and can contain duplicates
+	contextKey, tagsKey := cr.generateContextKey(metricSampleContext) // the generator will remove duplicates from cr.tagsBuffer (and doesn't mind the order)
 
 	if _, ok := cr.contextsByKey[contextKey]; !ok {
 		// making a copy of tags for the context since tagsBuffer
 		// will be reused later. This allow us to allocate one slice
 		// per context instead of one per sample.
```
Comment (djmitche) [inline pkg/aggregator/context_resolver.go:61]: I think this is inaccurate now?

File: pkg/aggregator/tags/store.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	return &Store{
+		tagsByKey: map[ckey.TagsKey]*Entry{},
+		enabled:   enabled,
+		telemetry: *newStoreTelemetry(name),
+	}
+}
+
+// Insert returns an Entry that corresponds to the key. If the key is
+// not in the cache, a new entry is stored in the Store with the tags
+// retrieved from the tagsBuffer.
```
Comment (djmitche) [inline pkg/aggregator/tags/store.go:54]: This will need some documentation about calling Release.

**Review by djmitche — APPROVED** [review]
Very nice!!

File: pkg/aggregator/tags/store.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	if r < 4 {
+		s.refsFreq[r-1]++
+	} else if r < 64 {
+		s.refsFreq[bits.Len64(r)]++ // Len(4) = 3, Len(63) = 6
+	} else {
+		s.refsFreq[7]++
+	}
+
+	n := len(e.tags)
+	if n < s.minSize {
```
Comment (ogaca-dd) [inline pkg/aggregator/tags/store.go:179]: As `s.minSize` is not initialized, the value is always `0` and `n < s.minSize` is always false.

File: pkg/aggregator/tags/store.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+var (
+	tlmHits          = newCounter("hits_total", "number of times cache already contained the tags")
+	tlmMiss          = newCounter("miss_total", "number of times cache did not contain the tags")
+	tlmEntries       = newGauge("entries", "number of entries in the tags cache")
+	tlmMaxEntries    = newGauge("max_entries", "maximum number of entries since last shrink")
+	tlmTagsetMinTags = newGauge("tagset_min_tags", "minimum number of tags in a tagset")
+	tlmTagsetMaxTags = newGauge("tagset_max_tags", "maximum number of tags in a tagset")
+	tlmTagsetSumTags = newGauge("tagset_sum_tags", "total number of tags stored in all tagsets by the cache")
+	tlmTagsetRefsCnt = newGauge("tagset_refs_count", "distribution of usage count of tagsets in the cache", "ge")
```
Comment (ogaca-dd) [inline pkg/aggregator/tags/store.go:144]: What is the meaning of `"ge"`?

File: pkg/aggregator/context_resolver_test.go [hunk 5]
```diff
@@ -17,9 +17,27 @@ import (
 	"github.com/stretchr/testify/require"
 
 	"github.com/DataDog/datadog-agent/pkg/aggregator/ckey"
+	"github.com/DataDog/datadog-agent/pkg/aggregator/tags"
 	"github.com/DataDog/datadog-agent/pkg/metrics"
 )
 
+// Helper functions to run tests and benchmarks for context resolver, time and check samplers.
+func testWithTagsStore(t *testing.T, test func(*testing.T, *tags.Store)) {
```
Comment (ogaca-dd) [inline pkg/aggregator/context_resolver_test.go:25]: 👍

**Review by ogaca-dd — APPROVED** [review]
LGTM, well written PR!

File: pkg/aggregator/tags/store.go (hunk 4 above)
Comment (vickenty) [reply pkg/aggregator/tags/store.go:144]: "greater or equal to", similar to "le" used by prometheus histograms.

File: pkg/aggregator/tags/store.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	enabled   bool
+	telemetry storeTelemetry
+}
+
+// NewStore returns new empty Store.
+func NewStore(enabled bool, name string) *Store {
+	return &Store{
+		tagsByKey: map[ckey.TagsKey]*Entry{},
+		enabled:   enabled,
+		telemetry: *newStoreTelemetry(name),
```
[the remark this answers is not in the document]
Comment (vickenty) [reply pkg/aggregator/tags/store.go]: Yes, but the constructor doesn't need to return a pointer. Fixed.

**Pull request merged**